controllers/clustertopology_controller.go

```diff
@@ -0,0 +1,177 @@
+	}
+
+	r.restConfig = mgr.GetConfig()
+	r.externalTracker = external.ObjectTracker{
+		Controller: controller,
+	}
+	return nil
+}
+
+func (r *ClusterTopologyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Fetch the Cluster instance.
+	cluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Return early, if the Cluster does not use a managed topology.
+	// NOTE: This should be removed as soon as we start to support Clusters moving from managed <-> unmanaged.
+	if cluster.Spec.Topology == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Return early if the Cluster is paused.
+	if annotations.IsPaused(cluster, cluster) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: Add patching as soon as we define how to report managed topology state into conditions
+
+	// In case the object is deleted, the managed topology stops to reconcile;
+	// (the other controllers will take care of deletion).
+	if cluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, cluster)
+}
+
+// reconcile handles cluster reconciliation.
+func (r *ClusterTopologyReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	// Gets the ClusterClass and the referenced templates.
+	class, err := r.getClass(ctx, cluster)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error reading the ClusterClass")
+	}
+
+	// Gets the current state of the Cluster.
+	currentState, err := r.getCurrentState(ctx, cluster)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error reading current state of the Cluster topology")
+	}
+
+	// Computes the desired state of the Cluster
+	desiredState, err := r.computeDesiredState(ctx, class, currentState)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error computing the desired state of the Cluster topology")
+	}
+
+	// Reconciles current and desired state of the Cluster
+	if err := r.reconcileState(ctx, currentState, desiredState); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error reconciling the Cluster topology")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// clusterClassToCluster is a handler.ToRequestsFunc to be used to enqueue requests for reconciliation
+// for Cluster to update when its own ClusterClass gets updated.
+func (r *ClusterTopologyReconciler) clusterClassToCluster(o client.Object) []ctrl.Request {
+	_, ok := o.(*clusterv1.ClusterClass)
+	if !ok {
+		panic(fmt.Sprintf("Expected a ClusterClass but got a %T", o))
+	}
+
+	// TODO: use the custom cache index for Cluster.Topology.Class to retry all the Cluster using a given ClusterClass
```

Comment:
This what I'm advocating in https://github.com/kubernetes-sigs/cluster-api/pull/4967#issuecomment-883377554; but I'm fine as well it this will be separated PR or if it will be included in this one depending on what merges first.

Reply:
I'm also fine with both variants. Just wanted to make sure I got it right / we're all on the same page.

Reply:
Yes, the ClusterClass index is being created in https://github.com/kubernetes-sigs/cluster-api/pull/4967. 
As @fabriziopandini suggested I will wait for this PR to merge before we merge the index PR. This will also give us time to have consensus on the package for indexes discussion.